stdlib/functional.makam

```diff
@@ -86,7 +96,15 @@ hoist_lifts FunctionalCode LogicCode :-
 do : A -> A -> prop.
 
 do FunctionalCode Result :-
+  or (dyn.eq FunctionalCode (lift (_: A -> prop)))
+     (dyn.eq FunctionalCode (lift (_: prop))),
   hoist_lifts FunctionalCode LogicCode,
   LogicCode Result.
 
+(* version of `functional.do` when no result is produced *)
+do : unit -> prop.
```

Comment:
Is it possible to change this to be `A`? I'm getting the following type error:

```prolog
functional.do (is_equal !(plus Length 1) OutputType1 OutputType2).
```
```
$ makam test/init --run-tests

	Makam, version 0.7.36


!! Error in file ./test/../src/ast/core/semantics.makam:436:24-32:
   Variable is_equal with type A -> fathom.ast.core.semantics.value -> fathom.ast.core.semantics.value -> unit does not exist.
```

Reply:
Ahhh! Interesting - yeah I think I was confused because I was reading `!` as 'use the last parameter as the output' when it's in `functional.do` - is there a better way to look at it?

Reply:
Hmm, even with the `!` around the outside I get:
```
!! Error in file ./test/../src/ast/core/semantics.makam:537:25-33:
   Variable is_equal with type int -> fathom.ast.core.semantics.value -> fathom.ast.core.semantics.value -> unit -> prop does not exist.
```

Reply:
Hm, that's weird. Could you point me to the commit to see what's going on?

The way I'm thinking of `!` is that it turns a predicate (either nonary or unary) into a computation yielding a certain result; and computations are what can get executed through `functional.do`. I have a branch that makes that more explicit in the types, but requires syntactic sugar.